src/python/pants/engine/target.py

```diff
@@ -915,12 +916,56 @@ def __init__(
         bulleted_list_sep = "\n  * "
         super().__init__(
             f"Multiple of the registered implementations for {goal_description} work for "
-            f"{target.address} (target type {repr(target.alias)}).\n\n"
-            "It is ambiguous which implementation to use. Possible implementations:"
+            f"{target.address} (target type {repr(target.alias)}). It is ambiguous which "
+            "implementation to use.\n\nPossible implementations:"
             f"{bulleted_list_sep}{bulleted_list_sep.join(possible_config_types)}"
         )
 
 
+class AmbiguousCodegenImplementationsException(Exception):
+    """Exception for when there are multiple codegen implementations and it is ambiguous which to
+    use."""
+
+    def __init__(
+        self,
+        generators: Iterable[Type["GenerateSourcesRequest"]],
+        *,
+        for_sources_types: Iterable[Type["Sources"]],
+    ) -> None:
+        bulleted_list_sep = "\n  * "
+        all_same_generator_paths = (
+            len(set((generator.input, generator.output) for generator in generators)) == 1
+        )
+        example_generator = list(generators)[0]
+        input = example_generator.input.__name__
+        if all_same_generator_paths:
+            output = example_generator.output.__name__
+            possible_generators = sorted(generator.__name__ for generator in generators)
+            super().__init__(
+                f"Multiple of the registered code generators can generate {output} from {input}. "
+                "It is ambiguous which implementation to use.\n\nPossible implementations:"
+                f"{bulleted_list_sep}{bulleted_list_sep.join(possible_generators)}"
+            )
+        else:
+            possible_output_types = sorted(
+                generator.output.__name__
+                for generator in generators
+                if issubclass(generator.output, tuple(for_sources_types))
+            )
+            possible_generators_with_output = [
+                f"{generator.__name__} -> {generator.output.__name__}"
+                for generator in sorted(generators, key=lambda generator: generator.output.__name__)
+            ]
+            super().__init__(
+                f"Multiple of the registered code generators can generate one of "
+                f"{possible_output_types} from {input}. It is ambiguous which implementation to "
+                f"use. This can happen when the call site requests too many different output types "
+                f"from the same original protocol sources.\n\nPossible implementations with their "
+                f"output type: {bulleted_list_sep}"
+                f"{bulleted_list_sep.join(possible_generators_with_output)}"
```

Comment:
This reads:

```
Multiple of the registered code generators can generate one of ['FortranSources', 'SmalltalkSources'] from AvroSources. It is ambiguous which implementation to use. This can happen when the call site requests too many different output types from the same original protocol sources.

Possible implementations with their output type:
  * FortranGenerator1 -> FortranSources
  * SmalltalkGenerator -> SmalltalkSources
```